### PR TITLE
Support `findAndModify`

### LINKED ISSUE #164: Support `findAndModify`
Improve support for `findAndModify` command with the following arguments:
* `findAndModify` (collection name)
* `query`
* `sort`
* `remove`
* `update`
* `new`
* `upsert`

Ignore `bypassDocumentValidation`, `writeConcern`, `maxTimeMS`, `collation`, `hint`, `comment` arguments.
Fail if `arrayFilters`, `let`, `fields` argument is present.

We have a basic implementation already, but it misses support for many arguments and does not have any tests. We should add them.

### PR DESCRIPTION
**Pull request by w84thesun**
Closes #164.

### CODE REVIEW

File: integration/findandmodify_test.go [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+		"EmptyQueryRemove": {
+			command: bson.D{
+				{"findAndModify", collection.Name()},
+				{"query", bson.D{}},
+				{"remove", true},
+			},
+			response: bson.D{
+				{"lastErrorObject", bson.D{{"n", int32(1)}}},
+				{
+					"value",
```
Comment (seeforschauer) [inline integration/findandmodify_test.go:45]: can this doc representation be somehow similar to the one on line 57?

File: internal/handlers/pg/msg_delete.go [hunk 2]
```diff
@@ -142,3 +129,25 @@ func (h *Handler) MsgDelete(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg,
...
+		ids[i] = must.NotFail(fjson.Marshal(id))
+	}
+
+	sql := fmt.Sprintf(
+		"DELETE FROM %s WHERE _jsonb->'_id' IN (%s)",
+		pgx.Identifier{db, collection}.Sanitize(), strings.Join(placeholders, ", "),
+	)
+	tag, err := h.pgPool.Exec(ctx, sql, ids...)
+	if err != nil {
+		// TODO check error code
```
Comment (seeforschauer) [inline internal/handlers/pg/msg_delete.go:149]: can it be done in this PR?

File: integration/findandmodify_test.go (hunk 1 above)
Comment (w84thesun) [reply integration/findandmodify_test.go:45]: I get it right you want to extract something from response. Looks like it's not an option as there would be difference in `lastErrorObject` between those 2 tests.

File: internal/handlers/pg/msg_delete.go (hunk 2 above)
Comment (w84thesun) [reply internal/handlers/pg/msg_delete.go:149]: It's doesn't looks like something related to this PR.

File: internal/handlers/pg/msg_insert.go [hunk 3]
```diff
@@ -94,3 +88,18 @@ func (h *Handler) MsgInsert(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg,
 
 	return &reply, nil
 }
+
+func (h *Handler) insert(ctx context.Context, doc any, db string, collection string) error {
+	d := doc.(*types.Document)
+	sql := fmt.Sprintf("INSERT INTO %s (_jsonb) VALUES ($1)", pgx.Identifier{db, collection}.Sanitize())
+	b, err := fjson.Marshal(d)
+	if err != nil {
+		return err
```
Comment (seeforschauer) [inline internal/handlers/pg/msg_insert.go:97]:
```suggestion
		return lazyerrors.Error(err)
```

File: internal/handlers/pg/msg_insert.go [hunk 4]
```diff
@@ -94,3 +88,18 @@ func (h *Handler) MsgInsert(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg,
...
+	}
+
+	if _, err = h.pgPool.Exec(ctx, sql, b); err != nil {
+		return err
```
Comment (seeforschauer) [inline internal/handlers/pg/msg_insert.go:101]:
```suggestion
		return lazyerrors.Error(err)
```

File: internal/handlers/pg/msg_update.go [hunk 5]
```diff
@@ -187,3 +186,13 @@ func (h *Handler) MsgUpdate(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg,
 
 	return &reply, nil
 }
+
+func (h *Handler) update(ctx context.Context, db string, collection string, doc *types.Document) (pgconn.CommandTag, error) {
+	sql := fmt.Sprintf("UPDATE %s SET _jsonb = $1 WHERE _jsonb->'_id' = $2", pgx.Identifier{db, collection}.Sanitize())
+	id := must.NotFail(doc.Get("_id"))
+	tag, err := h.pgPool.Exec(ctx, sql, must.NotFail(fjson.Marshal(doc)), must.NotFail(fjson.Marshal(id)))
+	if err != nil {
+		return nil, err
```
Comment (seeforschauer) [inline internal/handlers/pg/msg_update.go:195]:
```suggestion
		return nil, lazyerrors.Error(err)
```

File: internal/handlers/pg/msg_findandmodify.go [hunk 6]
```diff
@@ -110,23 +87,144 @@ func (h *Handler) MsgFindAndModify(ctx context.Context, msg *wire.OpMsg) (*wire.
...
+	}
+
+	var sort *types.Document
+	if sort, err = common.GetOptionalParam(document, "sort", sort); err != nil {
+		return nil, err
+	}
+
+	var update *types.Document
+	updateParam, err := document.Get("update")
+	if err != nil && !remove {
```
Comment (seeforschauer) [inline internal/handlers/pg/msg_findandmodify.go:206]: if err != nil 
or
update && remove

isn't it? Could you please add test for it?

File: internal/handlers/common/params.go [hunk 7]
```diff
@@ -50,13 +50,38 @@ func GetOptionalParam[T types.Type](doc *types.Document, key string, defaultValu
...
+	case float64:
+		return v != 0, nil
+	case bool:
+		return v, nil
+	case int32:
+		return v != 0, nil
+	case int64:
+		return v != 0, nil
+	default:
+		msg := fmt.Sprintf(`BSON field '%s' is the wrong type '%s', expected type 'bool'`, key, AliasFromType(v))
```
Comment (seeforschauer) [inline internal/handlers/common/params.go:80]: <string value> "true" | "false" will fail, is it ok?

Comment (w84thesun) [reply internal/handlers/common/params.go:80]: I don't get it. What do you mean?

File: internal/handlers/pg/msg_findandmodify.go (hunk 6 above)
Comment (w84thesun) [reply internal/handlers/pg/msg_findandmodify.go:206]: There is a test for that → "NotEnoughParameters".

File: internal/handlers/common/update.go [hunk 8]
```diff
@@ -48,3 +50,40 @@ func UpdateDocument(doc, update *types.Document) error {
 
 	return nil
 }
+
+// updateOperator represents update operators aliases.
+type updateOperator int8
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply internal/handlers/common/update.go:55]: It unifies the way we use them and makes `HasUpdateOperator` easier.

File: internal/handlers/common/params.go (hunk 7 above)
Comment (seeforschauer) [reply internal/handlers/common/params.go:80]: I mean string value will fail.

**Review by seeforschauer — CHANGES_REQUESTED** [review]
LGTM, except the string value

File: internal/handlers/common/params.go (hunk 7 above)
Comment (w84thesun) [reply internal/handlers/common/params.go:80]: String values should fail. There is a test for string values passed for boolean fields, like `BadRemoveType`.

Comment (seeforschauer) [reply internal/handlers/common/params.go:80]: That's the question. Here we can extend FerretDB friendliness to the user without being incompatible in responses. 
What are our thoughts here?

Comment (AlekSi) [reply internal/handlers/common/params.go:80]: We don't do that

Comment (seeforschauer) [reply internal/handlers/common/params.go:80]: ok

**Review by seeforschauer — APPROVED** [review]

File: internal/handlers/common/params.go [hunk 9]
```diff
@@ -50,13 +50,38 @@ func GetOptionalParam[T types.Type](doc *types.Document, key string, defaultValu
...
+// Non-zero value for double, long and int return true.
+func GetBoolOptionalParam(doc *types.Document, key string) (bool, error) {
+	v, err := doc.Get(key)
+	if err != nil {
+		return false, nil
+	}
+
+	switch v := v.(type) {
+	case float64:
+		return v != 0, nil
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply internal/handlers/common/params.go:73]: I've added test for that case.

**Review by AlekSi — COMMENTED** [review]
All new functions and methods should have documentation comments.

And please resolve `lll`'s warning

**Comment by w84thesun** [comment]
> All new functions and methods should have documentation comments.
> 
> And please resolve `lll`'s warning

Done.

File: internal/handlers/pg/msg_update.go [hunk 10]
```diff
@@ -187,3 +186,14 @@ func (h *Handler) MsgUpdate(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg,
 
 	return &reply, nil
 }
+
+// delete prepares and executes actual DELETE request to Postgres.
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply internal/handlers/pg/msg_update.go]: Fixed

File: internal/handlers/pg/msg_insert.go [hunk 11]
```diff
@@ -95,3 +89,19 @@ func (h *Handler) MsgInsert(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg,
 
 	return &reply, nil
 }
+
+// delete prepares and executes actual INSERT request to Postgres.
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply internal/handlers/pg/msg_insert.go]: Fixed

File: internal/handlers/pg/msg_findandmodify.go [hunk 12]
```diff
@@ -110,23 +85,150 @@ func (h *Handler) MsgFindAndModify(ctx context.Context, msg *wire.OpMsg) (*wire.
...
+		err = common.UpdateDocument(upsert, params.update)
+		if err != nil {
+			return nil, err
+		}
+
+		_, err = h.update(ctx, params.db, params.collection, upsert)
+		if err != nil {
+			return nil, err
+		}
+	} else {
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply internal/handlers/pg/msg_findandmodify.go:135]: We replacing document with the new one.

File: internal/handlers/pg/msg_findandmodify.go [hunk 13]
```diff
@@ -110,23 +85,150 @@ func (h *Handler) MsgFindAndModify(ctx context.Context, msg *wire.OpMsg) (*wire.
...
+			Documents: []*types.Document{must.NotFail(types.NewDocument(
+				"lastErrorObject", must.NotFail(types.NewDocument("n", int32(1))),
+				"value", resDocs[0],
+				"ok", float64(1),
+			))},
+		}))
+		return &reply, nil
+	}
+
+	if common.HasUpdateOperator(params.update) { //nolint:nestif // no way to make it simple now
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply internal/handlers/pg/msg_findandmodify.go]: Done.

**Review by seeforschauer — APPROVED** [review]

**Comment by seeforschauer** [comment]
some checks (i.e. inclusion/exclusion) of user input are in `$elemMatch` PR #551.

**Review by seeforschauer — CHANGES_REQUESTED** [review]
Would you please add more tests, 
- Cannot specify both an update and remove=true
- Cannot specify both upsert=true and remove=true 
- Cannot specify both new=true and remove=true; 'remove' always returns the deleted  document
- Cannot specify arrayFilters and remove=true
- Cannot specify arrayFilters and a pipeline update

Thank you.

**Comment by w84thesun** [comment]
Thanks, I've missed those checks.

> * Cannot specify arrayFilters and remove=true
> * Cannot specify arrayFilters and a pipeline update

I don't sure we support `arrayFilters` right now. Although we definitely don't support pipeline updates.

I've added other 3 tests.

**Comment by seeforschauer** [comment]
Thank you!

> I don't sure we support arrayFilters right now. Although we definitely don't support pipeline updates.


Maybe mention somewhere in the issue description that the `findAndModify` is supported w/o `arrayFilters` and pipeline updates?

**Comment by w84thesun** [comment]
> Thank you!
> 
> > I don't sure we support arrayFilters right now. Although we definitely don't support pipeline updates.
> 
> Maybe mention somewhere in the issue description that the `findAndModify` is supported w/o `arrayFilters` and pipeline updates?

`arrayFilters` are already mentioned in the issue and marked as unimplemented.

File: internal/handlers/pg/msg_findandmodify.go [hunk 14]
```diff
@@ -35,27 +32,14 @@ func (h *Handler) MsgFindAndModify(ctx context.Context, msg *wire.OpMsg) (*wire.
 		return nil, lazyerrors.Error(err)
 	}
 
-	// TODO https://github.com/FerretDB/FerretDB/issues/164
-
 	unimplementedFields := []string{
-		"sort",
-		"update",
 		"arrayFilters",
 		"let",
```
Comment (seeforschauer) [inline internal/handlers/pg/msg_findandmodify.go:37]: could you please add here a `field` field to make sure it fails if present?

File: internal/handlers/pg/msg_findandmodify.go [hunk 15]
```diff
@@ -35,27 +32,14 @@ func (h *Handler) MsgFindAndModify(ctx context.Context, msg *wire.OpMsg) (*wire.
...
-		if err := common.UnimplementedNonDefault(document, field, func(v any) bool {
-			b, ok := v.(bool)
-			return ok && !b
-		}); err != nil {
-			return nil, err
-		}
-	}
-
 	ignoredFields := []string{
 		"fields",
```
Comment (seeforschauer) [inline internal/handlers/pg/msg_findandmodify.go:44]: and remove `fields` here.

File: internal/handlers/pg/msg_findandmodify.go (hunk 14 above)
Comment (w84thesun) [reply internal/handlers/pg/msg_findandmodify.go:37]: Done.

File: internal/handlers/pg/msg_findandmodify.go (hunk 15 above)
Comment (w84thesun) [reply internal/handlers/pg/msg_findandmodify.go]: Done.

**Review by seeforschauer — APPROVED** [review]

**Review by AlekSi — APPROVED** [review]

**Pull request merged**